lib/lhm/throttler/slave_lag.rb

```diff
@@ -43,55 +45,83 @@ def throttle_seconds
+      def slaves
+        @slaves[@connection] ||= get_slaves
       end
 
       def get_slaves
-        @connection.select_values(SQL_SELECT_SLAVE_HOSTS)
+        slaves = []
+        slave_hosts = master_slave_hosts
+        while slave_hosts.any? do
+          host = slave_hosts.pop
+          slave = Slave.new(host, @get_config)
+          if slaves.map(&:host).exclude?(host) && slave.connection
+            slaves << slave
+            slave_hosts << slave.slave_hosts
+            slave_hosts.flatten!
+          end
+        end
+        slaves
+      end
+
+      def master_slave_hosts
+        Throttler.format_hosts(@connection.select_values(Slave::SQL_SELECT_SLAVE_HOSTS))
       end
 
       def max_current_slave_lag
-        max = slave_hosts.map { |slave| slave_lag(slave) }.flatten.push(0).max
+        max = slaves.map { |slave| slave.lag }.flatten.push(0).max
         Lhm.logger.info "Max current slave lag: #{max}"
         max
       end
+    end
+
+    class Slave
+      SQL_SELECT_SLAVE_HOSTS = "SELECT host FROM information_schema.processlist WHERE command='Binlog Dump'"
+      SQL_SELECT_MAX_SLAVE_LAG = 'SHOW SLAVE STATUS'
+
+      attr_reader :host, :connection
+
+      def initialize(host, get_config=nil)
+        @host = host
+        @connection = client(config(get_config))
+      end
 
-      def slave_lag(slave)
-        conn = slave_connection(slave)
-        if conn.respond_to?(:exec_query)
-          result = conn.exec_query(SQL_SELECT_MAX_SLAVE_LAG)
-          result.map { |row| row['Seconds_Behind_Master'].to_i }
-        else
-          result = conn.execute(SQL_SELECT_MAX_SLAVE_LAG)
-          fetch_slave_seconds(result)
-        end
-      rescue Error => e
-        raise Lhm::Error, "Unable to connect and/or query slave to determine slave lag. Migration aborting because of: #{e}"
+      def slave_hosts
+        Throttler.format_hosts(query_connection(SQL_SELECT_SLAVE_HOSTS, 'host'))
       end
 
-      def slave_connection(slave)
-        adapter_method = defined?(Mysql2) ? 'mysql2_connection' : 'mysql_connection'
-        config = ActiveRecord::Base.connection_pool.spec.config.dup
-        config[:host] = slave
-        ActiveRecord::Base.send(adapter_method, config)
+      def lag
+        query_connection(SQL_SELECT_MAX_SLAVE_LAG, 'Seconds_Behind_Master')
       end
 
-      # This method fetch the Seconds_Behind_Master, when exec_query is no available, on AR 2.3.
-      def fetch_slave_seconds(result)
-        unless result.is_a? Mysql::Result
-          Lhm.logger.info "Not a Mysql::Result from the slave assuming 0 lag"
-          return 0
+      private
+
+      def client(config)
+        begin
+          Mysql2::Client.new(config)
+        rescue Mysql2::Error => e
+          Lhm.logger.info "Error conecting to #{@host}: #{e}"
+          nil
         end
+      end
 
-        keys = []
-        result.each_hash { |h| keys << h['Seconds_Behind_Master'].to_i }
-        keys
+      def config(get_config)
```

Comment:
Does passing in a config hash to LHM makes more sense than passing in a connection? It seems like that the consumer of this API should already have a connection object handy and that might be more appropriate to pass in rather than have this class create a new connection object?